consensus/src/experimental/buffer_manager.rs

```diff
@@ -290,6 +365,117 @@ impl StateManager {
+            while current_cursor.is_some() {
+                // update signature
+                let buffer_item = take_elem(&current_cursor);
+                if !buffer_item.has_been_executed() {
+                    // a reset has happened
+                    // we do nothing except put it back
+                    set_elem(&current_cursor, buffer_item);
+                    break;
+                }
+
+                if buffer_item.get_commit_info() == commit_ledger_info.commit_info() {
+                    // it is possible that we already signed this buffer item (double check after the final integration)
+                    if matches!(buffer_item, BufferItem::Executed(_)) {
+                        // we have found the buffer item
+                        let (signed_buffer_item, commit_vote) =
+                            buffer_item.advance_to_signed(self.author, sig.clone(), &self.verifier);
+
+                        set_elem(&current_cursor, signed_buffer_item);
+
+                        // send out commit vote
+                        self.commit_msg_tx
+                            .broadcast(ConsensusMsg::CommitVoteMsg(Box::new(commit_vote)))
+                            .await;
+
+                        self.try_advance_signing_root();
+                    }
+                    break;
+                }
+                current_cursor = get_next(&current_cursor);
+            }
+        }
+        // otherwise, a reset happened, we do nothing
+        Ok(())
+    }
+
+    async fn process_signing_response(
+        &mut self,
+        signing_resp: SigningResponse,
+    ) -> anyhow::Result<()> {
+        let SigningResponse {
+            signature_result,
+            commit_ledger_info,
+        } = signing_resp;
+        if let Ok(sig) = signature_result {
+            self.process_successful_signing_resp(sig, commit_ledger_info)
+                .await
+        } else {
+            // try next signature if signing failure
+            // note that we are not retrying exactly the failed sig
+            // the failed sig will be re-tried in the future, unless a reset happens
+            let current_cursor = self.signing_root.clone();
+            if current_cursor.is_some() {
+                let buffer_item = get_elem(&current_cursor);
+                if let BufferItem::Executed(executed) = &(*buffer_item) {
+                    self.signing_phase_tx
+                        .send(SigningRequest {
+                            ordered_ledger_info: executed.ordered_proof.clone(),
+                            commit_ledger_info: executed.generate_commit_ledger_info(),
+                        })
+                        .await?;
+                }
+            }
+            Ok(())
+        }
+    }
+
+    async fn process_commit_msg(&mut self, commit_msg: VerifiedEvent) -> anyhow::Result<()> {
+        match commit_msg {
+            VerifiedEvent::CommitVote(cv) => {
+                // travel the whole buffer
```

Comment:
hashmap seems better here

Reply:
we can add a hashmap for a quick lookup. I think the buffer itself still should be a linkedlist.

Reply:
Added hashmap to TODO.